### PR TITLE
SPARKC-73 support cross cluster table join and write for Spark SQL

### PR DESCRIPTION
**Pull request by alexliu68**

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSQLContext.scala [hunk 1]
```diff
@@ -37,6 +42,153 @@ class CassandraSQLContext(sc: SparkContext) extends SQLContext(sc) {
...
+                      tableWriteConf.get(Seq(table, keyspace, c)).getOrElse(
+                        keyspaceWriteConf.get(Seq(keyspace, c)).getOrElse(
+                          clusterWriteConf.get(c).getOrElse(WriteConf.fromSparkConf(conf))))
+      case _       => tableWriteConf.get(Seq(table, keyspace)).getOrElse(
+                        keyspaceWriteConf.get(Seq(keyspace)).getOrElse(WriteConf.fromSparkConf(conf)))
+    }
+  }
+
+  @transient
+  private val clusterCassandraConnConf = mutable.Map[String, SparkConf]()
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSQLContext.scala]: Why not `CassandraConnectorConf` as a value?

Comment (alexliu68) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSQLContext.scala]: miss it

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala [hunk 2]
```diff
@@ -16,48 +16,151 @@ trait EmbeddedCassandra {
...
-  // TODO change to CASSANDRA_HOST
-  val HostProperty = "IT_TEST_CASSANDRA_HOST"
+  private[connector] var cassandraRunners: IndexedSeq[Option[CassandraRunner]] = IndexedSeq(None)
+
+  private[connector] var currentConfigTemplates: IndexedSeq[String] = IndexedSeq()
+
+  def getProps(index: Integer): Map[String, String] = {
+    require(hosts.isEmpty || index < hosts.size, s"$index index is overflow the size of ${hosts.size}")
+    val host = getHost(index).getHostAddress
+    Map("seeds"               -> host,
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: I think this should be configurable somewhere - maybe not for this ticket but in the future i would be nice not to only start nodes in separate clusters but also in a single dc.

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala [hunk 3]
```diff
@@ -16,48 +16,151 @@ trait EmbeddedCassandra {
...
 
+abstract class NodeProperty
+
+case class HostProperty(propertyName: String) extends NodeProperty
+case class NativePortProperty(propertyName: String) extends NodeProperty
+case class RpcPortProperty(propertyName: String) extends NodeProperty
+
+object UserDefinedProperty {
+  type propertyType = NodeProperty {val propertyName: String}
+  final val hostProperty = HostProperty("IT_TEST_CASSANDRA_HOSTS")
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: Well I think they can be case objects

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala [hunk 4]
```diff
@@ -16,48 +16,151 @@ trait EmbeddedCassandra {
...
 }
 
+abstract class NodeProperty
+
+case class HostProperty(propertyName: String) extends NodeProperty
+case class NativePortProperty(propertyName: String) extends NodeProperty
+case class RpcPortProperty(propertyName: String) extends NodeProperty
+
+object UserDefinedProperty {
+  type propertyType = NodeProperty {val propertyName: String}
```
Comment (jacek-lewandowski) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: I think the name of this type should start with uppercase

**Comment by jacek-lewandowski** [comment]
Reviewed

**Comment by jacek-lewandowski** [comment]
Only minor things, other than that +1

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala (hunk 3 above)
Comment (alexliu68) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: agree

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala (hunk 4 above)
Comment (alexliu68) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: changed

File: spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala (hunk 2 above)
Comment (alexliu68) [inline spark-cassandra-connector-embedded/src/main/scala/com/datastax/spark/connector/embedded/EmbeddedCassandra.scala]: I will open another ticket for it

**Pull request merged**